src/context/provided_types.rs

```diff
@@ -0,0 +1,92 @@
+use std::{marker::PhantomData, mem, sync::OnceLock};
+
+use better_any::TidAble;
+
+use crate::FileRunContext;
+
+pub trait FromFileRunContextInstanceProvider<'a>: Sized {
+    type Parent: FromFileRunContextInstanceProviderFactory<Provider<'a> = Self>;
```

Comment:
I tried figuring out this "hey Rust I assure you the parent and child are 1-to-1 "circularly" matching types" but what I came up with didn't convince it, but then saw someone explaining this way of doing it (which worked)

Reply:
What's the parent/child relationship here? I'm not really following.

Reply:
Right "parent"/"child" isn't the most helpful description, it's "factory"/"type produced by that factory" (where here those are "instance provider factory"/"instance-provider instance")